File: examples/async_auto_fix.rs
#[derive(Debug, PartialEq, PartialOrd)]
struct NumberItem {
    number: i32,
}

impl std::fmt::Display for NumberItem {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.number.fmt(f)
    }
}

// The item container lets a type that is normally not debuggable, displayable,
// or sortable to be. This is only really useful for graphical user interfaces.
impl openchecks::Item for NumberItem {
    type Value<'a>
        = i32
    where
        Self: 'a;

    fn value(&self) -> Self::Value<'_> {
        self.number
    }
}

impl NumberItem {
    pub fn new(number: i32) -> Self {
        Self { number }
    }
}

struct IsZeroCheck {
    number: i32,
}

impl openchecks::CheckMetadata for IsZeroCheck {
    fn title(&self) -> std::borrow::Cow<str> {
        "Is Zero Check".into()
    }

    fn description(&self) -> std::borrow::Cow<str> {
        "Check if the number is zero".into()
    }
}

#[async_trait::async_trait]
impl openchecks::AsyncCheck for IsZeroCheck {
    type Item = NumberItem;

    type Items = Vec<Self::Item>;

    async fn async_check(&self) -> openchecks::CheckResult<Self::Item, Self::Items> {
        // In this case, we can automatically fix the error. If the number is
        // not zero, set it to zero!
        let can_fix = true;
        let can_skip = false;

        // The items argument doesn't need to have any items, unless that
        // provides useful context. Generally this is reserved for failures.
        if self.number == 0 {
            openchecks::CheckResult::new_passed("The number is zero.", None, can_fix, can_skip)
        } else {
            openchecks::CheckResult::new_failed(
                "The number is not zero.",
                Some(vec![NumberItem::new(self.number)]),
                can_fix,
                can_skip,
            )
        }
    }

    async fn async_auto_fix(&mut self) -> Result<(), openchecks::Error> {
        self.number = 0;

        Ok(())
    }
}

impl IsZeroCheck {
    pub fn new(number: i32) -> Self {
        Self { number }
    }
}

#[tokio::main]
async fn main() {
    let mut check = IsZeroCheck::new(1);
    let result_fail = openchecks::async_run(&check).await;

    assert!(result_fail.status().has_failed());
    assert!(result_fail.can_fix());

    // When an auto-fix is ran, then the run function is automatically called
    // after to validate that the auto-fix fixed the issue or not.
    let result_pass = openchecks::async_auto_fix(&mut check).await;

    assert!(!result_pass.status().has_failed());
}

File: bindings/python/examples/async_auto_fix.py
"""This is an example of running a check that can automatically be fixed.

The check requires the same methods as the simple async check example, but also
include an ``async_auto_fix`` method.

The ``async_auto_fix`` method does not need to do anything other than attempt
to fix the issue, as the ``async_auto_fix`` function (similar to the
``async_run`` function) will automatically run the ``async_run`` function after
attempting the fix and returning the result. The rule with exceptions in the
``async_check`` method is the same as the ``async_auto_fix`` method. If there's
an exception raised, then the ``async_auto_fix`` function will return a
``SystemError`` status.
"""

# ruff: noqa: D101,D102,D103,D107,S101

import asyncio

import openchecks


class IsZeroCheck(openchecks.AsyncBaseCheck):
    def __init__(self, number: int) -> None:
        self.__number = number
        super().__init__()

    def title(self) -> str:
        return "Is Zero Check"

    def description(self) -> str:
        return "Check if the number is zero"

    async def async_check(self) -> openchecks.CheckResult:
        if self.__number == 0:
            return openchecks.CheckResult.passed("The number is zero.")
        return openchecks.CheckResult.failed(
            "The number is not zero.",
            [openchecks.Item(self.__number)],
            can_fix=True,
        )

        # Note: The check method must return a CheckResult in all of its
        # branches. If it doesn't, then the system will return a `SystemError`,
        # and that cannot be recovered from unless the check is fixed.

    async def async_auto_fix(self) -> None:
        self.__number = 0


async def main() -> None:
    check = IsZeroCheck(1)
    result_fail = await openchecks.async_run(check)

    assert result_fail.status().has_failed()

    result_pass = await openchecks.async_auto_fix(check)

    assert not result_pass.status().has_failed()


if __name__ == "__main__":
    asyncio.run(main())
